TST: Table data type comparison fails with numpy-dev

These are new failures, probably from numpy-dev, which is now 1.21.dev after the release of 1.20.

One example (others are variants of such comparisons):

```
>       assert col.dtype == dtype_expected
E       AssertionError: assert dtype('O') == dtype('S1')
E        +  where dtype('O') = <MaskedColumn name='col0' dtype='object' length=8>\n 0\n 1\n 2\n 3\n 4\n 5\n 6\n--.dtype
```

Affected tests:

* `astropy/table/tests/test_masked.py::TestTableInit::test_init_from_sequence_data_numeric_untyped[shape0-S]`
* `astropy/table/tests/test_masked.py::TestTableInit::test_init_from_sequence_data_numeric_untyped[shape0-U]`
* `astropy/table/tests/test_masked.py::TestTableInit::test_init_from_sequence_data_numeric_untyped[shape1-S]`
* `astropy/table/tests/test_masked.py::TestTableInit::test_init_from_sequence_data_numeric_untyped[shape1-U]`
* `astropy/table/tests/test_masked.py::TestTableInit::test_init_from_sequence_data_numeric_untyped[shape2-S]`
* `astropy/table/tests/test_masked.py::TestTableInit::test_init_from_sequence_data_numeric_untyped[shape2-U]`
* `astropy/table/tests/test_pprint.py::test_html_escaping`
* `docs/table/masking.rst::masking.rst`

Is this a matter of updating the expected results, or is there some compatibility bug to be fixed -- @mhvk or @taldcroft ?

I think it has something to do with this warning that we are ignoring.

https://github.com/astropy/astropy/blob/74d445ceaa61b5d23241210367902fbd60c9782c/setup.cfg#L122

When I un-ignore the warning, I get this:

```
    from astropy.table import Table, MaskedColumn
astropy/table/__init__.py:4: in <module>
    from .column import Column, MaskedColumn, StringTruncateWarning, ColumnInfo
astropy/table/column.py:22: in <module>
    from ._column_mixins import _ColumnGetitemShim, _MaskedColumnGetitemShim
astropy/table/_column_mixins.pyx:46: in init astropy.table._column_mixins
    ctypedef class numpy.ndarray [object PyArrayObject]:
E   RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 80 from C header, got 88 from PyObject
```

Perhaps as a result of this warning, the `dtype` is `O` instead of `S` or `U`, as expected. But I still don't understand why the test pass if I move the test code outside of `astropy` repo.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1251,6 +1251,10 @@ astropy.table
   triggered swapping when native endianness was stored with explicit
   ``dtype`` code ``'<'`` (or ``'>'``) instead of ``'='``. [#11288, #11294]
 
+- Fixed a compatibility issue with numpy 1.21. Initializing a Table with a
+  column like ``['str', np.ma.masked]`` was failing in tests due to a change in
+  numpy. [#11364]
+
 - Fixed bug when validating the inputs to ``table.hstack``, ``table.vstack``,
   and ``table.dstack``. Previously, mistakenly calling ``table.hstack(t1, t2)``
   (instead of ``table.hstack([t1, t2]))`` would return ``t1`` instead of raising

--- a/astropy/table/column.py
+++ b/astropy/table/column.py
@@ -196,6 +196,10 @@ def _convert_sequence_data_to_array(data, dtype=None):
         # converted to an error (which can happen during pytest).
         warnings.filterwarnings('always', category=UserWarning,
                                 message='.*converting a masked element.*')
+        # FutureWarning in numpy 1.21. See https://github.com/astropy/astropy/issues/11291
+        # and https://github.com/numpy/numpy/issues/18425.
+        warnings.filterwarnings('always', category=FutureWarning,
+                                message='.*Promotion of numbers and bools to strings.*')
         try:
             np_data = np.array(data, dtype=dtype)
         except np.ma.MaskError:

